Source/GoBot/FeedStrategy/PostsAndContactsAlgorithm.swift

```diff
@@ -0,0 +1,367 @@
+        AND claimed_at < ?
+        ORDER BY messages.claimed_at DESC
+        LIMIT ? OFFSET ?;
+        """)
+
+        let bindings: [Binding?] = [
+            userId,
+            userId,
+            Date().millisecondsSince1970,
+            limit,
+            offset ?? 0
+        ]
+
+        let colKey = Expression<MessageIdentifier>("key")
+        return try qry.bind(bindings).prepareRowIterator().map { row -> MessageIdentifier in
+            try row.get(colKey)
+        }
+    }
+
+    func fetchKeyValues(connection: Connection, userId: Int64, limit: Int, offset: Int?) throws -> [KeyValue] {
+        let qry = try connection.prepare("""
+        SELECT messages.*,
+               posts.*,
+               contacts.*,
+               tangles.*,
+               messagekeys.*,
+               authors.*,
+               author_about.*,
+               contact_author.author AS contact_identifier,
+               EXISTS (SELECT 1 FROM post_blobs WHERE post_blobs.msg_ref == messages.msg_id) as has_blobs,
+               EXISTS (SELECT 1 FROM mention_feed WHERE mention_feed.msg_ref == messages.msg_id) as has_feed_mentions,
+               EXISTS (SELECT 1 FROM mention_message WHERE mention_message.msg_ref == messages.msg_id) as has_message_mentions,
+               (SELECT COUNT(*) FROM tangles WHERE root == messages.msg_id) as replies_count,
+               (SELECT GROUP_CONCAT(authors.author, ';') FROM tangles JOIN messages AS tangled_message ON tangled_message.msg_id == tangles.msg_ref JOIN authors ON authors.id == tangled_message.author_id WHERE tangles.root == messages.msg_id LIMIT 3) as replies
+        FROM messages
+        LEFT JOIN posts ON messages.msg_id == posts.msg_ref
+        LEFT JOIN contacts ON messages.msg_id == contacts.msg_ref
+        LEFT JOIN tangles ON tangles.msg_ref == messages.msg_id
+        JOIN messagekeys ON messagekeys.id == messages.msg_id
+        JOIN authors ON authors.id == messages.author_id
+        LEFT JOIN abouts AS author_about ON author_about.about_id == messages.author_id
+        LEFT JOIN authors AS contact_author ON contact_author.id == contacts.contact_id
+        LEFT JOIN abouts AS contact_about ON contact_about.about_id == contacts.contact_id
+        WHERE type IN ('post', 'contact')
+        AND is_decrypted == false
+        AND hidden == false
+        AND (type <> 'post' OR posts.is_root == true)
+        AND (type <> 'contact' OR (contact_about.about_id IS NOT NULL AND contact_author.author NOT IN (SELECT key FROM pubs)))
+        AND (authors.author IN (SELECT authors.author FROM contacts
+                               JOIN authors ON contacts.contact_id == authors.id
+                               WHERE contacts.author_id = ? AND contacts.state == 1)
+             OR authors.id == ?)
+        AND authors.author NOT IN (SELECT key FROM pubs)
+        AND claimed_at < ?
+        ORDER BY claimed_at DESC
+        LIMIT ? OFFSET ?;
```

Comment:
Wow this query is a beast! Nice work fitting this all into one statement!

Reply:
Added comments explaining what it is doing at each step